posthog/temporal/usage_reports/run_usage_reports.py

```diff
@@ -0,0 +1,197 @@
+
+    total_orgs = len(inputs.org_reports)
+    total_orgs_sent = 0
+
+    pha_client.capture(
+        "internal_billing_events",
+        "organization usage report starting",
+        {
+            "total_orgs": total_orgs,
+            "region": get_instance_region(),
+        },
+        groups={"instance": settings.SITE_URL},
+    )
+
+    for org_report in inputs.org_reports.values():
+        try:
+            organization_id = org_report.organization_id
+
+            full_report = _get_full_org_usage_report(org_report, instance_metadata)
+            full_report_dict = _get_full_org_usage_report_as_dict(full_report)
+
+            # First capture the events to PostHog
+            if not inputs.skip_capture_event:
+                try:
+                    at_date_str = at_date.isoformat() if at_date else None
+                    capture_report(
+                        pha_client=pha_client,
+                        organization_id=organization_id,
+                        full_report_dict=full_report_dict,
+                        at_date=at_date_str,
+                    )
+                except Exception as err:
+                    logger.exception(f"Error capturing report for organization {organization_id}: {err}")
+
+            # Then send the reports to billing through SQS (only if the producer is available)
+            if has_non_zero_usage(full_report) and producer:
+                try:
+                    _queue_report(producer, organization_id, full_report_dict)
+                    total_orgs_sent += 1
+                except Exception as err:
+                    logger.exception(f"Error queueing report for organization {organization_id}: {err}")
+        except Exception as loop_err:
+            logger.exception(f"Error processing organization report: {loop_err}")
+
+    pha_client.capture(
+        "internal_billing_events",
+        "organization usage report complete",
+        {
+            "total_orgs": total_orgs,
+            "period_start": period_start.isoformat(),
+            "period_end": period_end.isoformat(),
+            "total_orgs_sent": total_orgs_sent,
+            "region": get_instance_region(),
+        },
+        groups={"instance": settings.SITE_URL},
+    )
+
+
+@workflow.defn(name="run-usage-reports")
+class RunUsageReportsWorkflow(PostHogWorkflow):
+    @staticmethod
+    def parse_inputs(inputs: list[str]) -> RunUsageReportsInputs:
+        """Parse inputs from the management command CLI."""
+        loaded = json.loads(inputs[0])
+        return RunUsageReportsInputs(**loaded)
+
+    @workflow.run
+    async def run(self, inputs: RunUsageReportsInputs) -> str:
+        query_usage_reports_inputs = QueryUsageReportsInputs(
+            at=inputs.at,
+        )
+        query_usage_reports_result = await workflow.execute_activity(
+            query_usage_reports,
+            query_usage_reports_inputs,
+            start_to_close_timeout=timedelta(minutes=20),
+        )
```

Comment:
Because we have no heartbeat timeout here and the start_to_close timeout is big-ish, if the pod running this activity dies either from OOM, termination from host node, or due to a deployment, then temporal will wait 20-mins before declaring this activity as failed. Unsure on whether this is something that may interrupt your flow.

Reply:
The default `retry_policy` is "infinite" retries. So, if this fails for some reason, it will retry until it succeeds. if this isn't expected behavior, then adding a `RetryPolicy(...)` would be good

Reply:
Added but might also be good to add a heartbeat, need to learn how that works